ir/json_parser.h

```diff
@@ -56,35 +59,31 @@ class JsonString : public JsonData, public std::string {
     DECLARE_TYPEINFO(JsonString, JsonData);
 };
 
-class JsonVector : public JsonData, public std::vector<JsonData *> {
+class JsonVector : public JsonData, public std::vector<std::unique_ptr<JsonData>> {
  public:
     JsonVector() {}
-    JsonVector(const std::vector<JsonData *> &v)  // NOLINT(runtime/explicit)
-        : std::vector<JsonData *>(v) {}
-    JsonVector &operator=(const JsonVector &) & = default;
-    JsonVector &operator=(JsonVector &&) & = default;
+    JsonVector(std::vector<std::unique_ptr<JsonData>> &&v)  // NOLINT(runtime/explicit)
+        : std::vector<std::unique_ptr<JsonData>>(std::move(v)) {}
+    JsonVector &operator=(const JsonVector &) = delete;
+    JsonVector &operator=(JsonVector &&) = default;
 
     DECLARE_TYPEINFO(JsonVector, JsonData);
 };
 
-class JsonObject : public JsonData, public ordered_map<std::string, JsonData *> {
-    bool _hasSrcInfo = true;
-
+class JsonObject : public JsonData, public ordered_map<std::string, std::unique_ptr<JsonData>> {
  public:
     JsonObject() {}
-    JsonObject(const JsonObject &obj) = default;
-    JsonObject &operator=(JsonObject &&) & = default;
-    JsonObject(const ordered_map<std::string, JsonData *> &v)  // NOLINT(runtime/explicit)
-        : ordered_map<std::string, JsonData *>(v) {}
-    int get_id() const;
-    std::string get_type() const;
-    std::string get_filename() const;
-    std::string get_sourceFragment() const;
-    int get_line() const;
-    int get_column() const;
-    JsonObject get_sourceJson() const;
-    bool hasSrcInfo() { return _hasSrcInfo; }
-    void setSrcInfo(bool value) { _hasSrcInfo = value; }
+    JsonObject(const JsonObject &obj) = delete;
+    JsonObject &operator=(JsonObject &&) = default;
+    JsonObject(ordered_map<std::string, std::unique_ptr<JsonData>> &&v)  // NOLINT(runtime/explicit)
+        : ordered_map<std::string, std::unique_ptr<JsonData>>(std::move(v)) {}
+
+    // FIXME -- need to create a copy of the string_view to lookup.  With C++20, this
+    // should not be needed, except ordered_map still requires it as its internal
+    // map is weird and doesn't support a templated find (and it is not clear how to).
+    using ordered_map<std::string, std::unique_ptr<JsonData>>::find;
```

Comment:
Will it make sense to switch to `string_map`? It does support heterogenous lookup. This would put all the keys into cstring cache though, not sure if this is good or not, but likely their variety is pretty small, so we can tolerate this?

Reply:
Almost all the keys will end up being cstrings at some point anyways, so that is not a big deal.